azure-iot-device/azure/iot/device/provisioning/internal/request_response_provider.py

```diff
@@ -0,0 +1,80 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import json
+from ..models.registration_result import RegistrationResult, RegistrationState
+
+logger = logging.getLogger(__name__)
+
+
+class RequestResponseMachine(object):
+    def __init__(self, state_based_provider):
+
+        self._state_based_provider = state_based_provider
+        self._state_based_provider.on_state_based_provider_message_received = self._receive_response
+
+        self.on_response_received = None
+
+    def send_request(self, publish_action):
```

Comment:
>publish_action [](start = 27, length = 14)

The actions inside action.py are meant to be internal to the provider.  Callers should not be aware that the concept of an action exists.

Reply:
yes I was having questions regarding this.  so should the polling machine call with the topic and the message and qos (in subscribe cases) as 3 separate parameters..which means 3 params get passed down to request response... where should the action get created ? if mqtt transport takes post connect or pre disconnect actions then actions must at least get created in request response...or the other option is to pass on the same information as request class which gets translated into action at the request response layer. 

---
In reply to: [277396270](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277396270) [](ancestors = 277396270)